chain/pruned_block_dispatcher.go

```diff
@@ -0,0 +1,554 @@
+package chain
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/btcjson"
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/peer"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/neutrino/query"
+	"github.com/lightningnetwork/lnd/ticker"
+)
+
+const (
+	// defaultRefreshPeersInterval represents the default polling interval
+	// at which we attempt to refresh the set of known peers.
+	defaultRefreshPeersInterval = 5 * time.Second
+
+	// defaultPeerReadyTimeout is the default amount of time we'll wait for
+	// a query peer to be ready to receive incoming block requests. Peers
+	// cannot respond to requests until the version exchange is completed
+	// upon connection establishment.
+	defaultPeerReadyTimeout = 15 * time.Second
+
+	// requiredServices are the requires services we require any candidate
+	// peers to signal such that we can retrieve pruned blocks from them.
+	requiredServices = wire.SFNodeNetwork | wire.SFNodeWitness
+
+	// prunedNodeService is the service bit signaled by pruned nodes on the
+	// network.
+	prunedNodeService wire.ServiceFlag = 2 << 10
+)
+
+// queryPeer represents a Bitcoin network peer that we'll query for blocks.
+// The ready channel serves as a signal for us to know when we can be sending
+// queries to the peer. Any messages received from the peer are sent through the
+// msgsRecvd channel.
+type queryPeer struct {
+	*peer.Peer
+	ready     chan struct{}
+	msgsRecvd chan wire.Message
+	quit      chan struct{}
+}
+
+// signalUponDisconnect closes the peer's quit chan to signal it has
+// disconnected.
+func (p *queryPeer) signalUponDisconnect(f func()) {
+	go func() {
+		p.WaitForDisconnect()
+		close(p.quit)
+		f()
+	}()
+}
+
+// SubscribeRecvMsg adds a OnRead subscription to the peer. All bitcoin messages
+// received from this peer will be sent on the returned channel. A closure is
+// also returned, that should be called to cancel the subscription.
+//
+// NOTE: This method exists to satisfy the query.Peer interface.
+func (p *queryPeer) SubscribeRecvMsg() (<-chan wire.Message, func()) {
+	return p.msgsRecvd, func() {}
```

Comment:
The lifetime of the subscription is bound to the lifetime of the `queryPeer` struct and its corresponding `query.Worker` spawned by the `query.WorkManager`, so the closure returned only gets called after the peer disconnects.

The closure exists as part of the API because of how neutrino handles message subscriptions, as it needs to clean up some state internally. This isn't needed here, which is why an empty closure is returned.